File: 0015-bloom-filter-any/main.jl
using SHA

mutable struct BloomFilter
    size::Int
    number_hashes::Int
    salt::AbstractString
    bit_array::Array{Int, 1}

    function BloomFilter(size::Int, number_hashes::Int, salt::AbstractString="")
        new(size, number_hashes, salt, zeros(Int, size))
    end
end

function add!(bf::BloomFilter, element::AbstractString)::Nothing
    for i in 0:bf.number_hashes-1
        digest = bytes2hex(sha1(string(bf.salt, element, i)))
        index = parse(BigInt, digest, base=16) % bf.size

        bf.bit_array[index + 1] = 1 # array index start from 1 not 0
    end
end

function lookup(bf::BloomFilter, element::AbstractString)::Bool
    for i in 0:bf.number_hashes-1
        digest = bytes2hex(sha1(string(bf.salt, element, i)))
        index = parse(BigInt, digest, base=16) % bf.size

        if checkbounds(Bool, bf.bit_array, index + 1) && bf.bit_array[index + 1] == 0
            return false
        end
    end

    return true
end

function estimate_dataset_size(bf::BloomFilter)::AbstractFloat
    m = bf.size
    k = bf.number_hashes
    n = -(m / k) * log(1 - sum(bf.bit_array) / m)

    return n
end

function union(bf::BloomFilter, other::BloomFilter)::BloomFilter
    if bf.size != other.size || bf.number_hashes != other.number_hashes
        error("Both filters must have the same size and hash count")
    end

    result = BloomFilter(bf.size, bf.number_hashes)
    result.bit_array = [x | y for (x, y) in zip(bf.bit_array, other.bit_array)]

    # new_bit_array = Int[]
    # for (x, y) in zip(bf.bit_array, other.bit_array)
    #     push!(new_bit_array, x | y)
    # end
    # result.bit_array = new_bit_array

    return result
end

function intersection(bf::BloomFilter, other::BloomFilter)::BloomFilter
    if bf.size != other.size || bf.number_hashes != other.number_hashes
        error("Both filters must have the same size and hash count")
    end

    result = BloomFilter(bf.size, bf.number_hashes)
    result.bit_array = [x & y for (x, y) in zip(bf.bit_array, other.bit_array)]

    return result
end

function main()
    coffees = [
        "Iced Coffee",
        "Iced Coffee with Milk",
        "Espresso",
        "Espresso Macchiato",
        "Flat White",
        "Latte Macchiato",
        "Cappuccino",
        "Mocha",
    ]

    bloom = BloomFilter(20, 2)
    for drink in coffees
        add!(bloom, drink)
        println(bloom.bit_array)
    end

    println("---Experiment #1---")
    println(lookup(bloom, "Flat White"))
    println(lookup(bloom, "Americano"))
    println(estimate_dataset_size(bloom))

    #
    more_coffees = [
        "Iced Espresso",
        "Flat White",
        "Cappuccino",
        "Frappuccino",
        "Latte",
    ]

    bloom2 = BloomFilter(20, 2)
    for drink in more_coffees
        add!(bloom2, drink)
    end

    bloom3 = union(bloom2, bloom)
    println("---Experiment #2---")
    println(lookup(bloom3, "Mocha"))
    println(lookup(bloom3, "Frappuccino"))
    println(estimate_dataset_size(bloom3))

    bloom4 = intersection(bloom2, bloom)
    println("---Experiment #3---")
    println(lookup(bloom4, "Mocha"))
    println(lookup(bloom4, "Flat White"))
    println(estimate_dataset_size(bloom4))
end

main()

File: 0015-bloom-filter-any/rust/src/main.rs
use sha1::{Sha1, Digest /* for `new/0` from CoreWrapper<Sha1Core> */};
use num_bigint::BigInt;
use num_traits::cast::ToPrimitive;

// extern crate hex; // auto now

//
trait Sha1Hash {
    fn sha1_hash(&self, value: &str) -> String;
}

//
#[derive(Clone)]
struct BloomFilter {
    size: usize,
    number_hashes: i32,
    salt: String,
    bit_array: Vec<i32>,
}

impl BloomFilter {

    pub fn new(size: usize, number_hashes: i32) -> Self {
        Self::new_with_salt(size, number_hashes, "")
    }

    pub fn new_with_salt(size: usize, number_hashes: i32, salt: &str) -> Self {
        Self {
            size,
            number_hashes,
            salt: salt.to_string(),
            bit_array: vec![0 as i32; size],
        }
    }

    pub fn add(&mut self, element: &str) {
        for i in 0..self.number_hashes {
            let digest = self.sha1_hash(format!("{salt}{element}{i}", salt=self.salt).as_str());
            let index = BigInt::parse_bytes(digest.as_bytes(), 16).unwrap() % self.size;

            self.bit_array[index.to_usize().unwrap()] = 1;
        }
    }

    pub fn lookup(&self, element: &str) -> bool {
        for i in 0..self.number_hashes {
            let digest = self.sha1_hash(format!("{salt}{element}{i}", salt=self.salt).as_str());
            let index = BigInt::parse_bytes(digest.as_bytes(), 16).unwrap() % self.size;

            if self.bit_array[index.to_usize().unwrap()] == 0 {
                return false
            }
        }

        true
    }

    pub fn estimate_dataset_size(&self) -> f64 {
        let m = self.size as i32;
        let k = self.number_hashes;
        let n = -(m / k) as f64 * (1 as f64 - self.bit_array.iter().sum::<i32>() as f64 / m as f64).log(std::f64::consts::E);

        // println!("{n:.15}");
        n
    }

    pub fn union(&self, other: BloomFilter) -> Self {
        if self.size != other.size || self.number_hashes != other.number_hashes {
            panic!("Both filters must have the same size and hash count")
        }

        /*
            let zipped_bit_array = self.bit_array.iter().zip(other.bit_array.iter()).collect::<Vec<_>>();
            let zipped_bit_array = self.bit_array.iter().zip(other.bit_array.iter()).collect::<Vec<(&i32, &i32)>>();
            let zipped_bit_array: [(&i32, &i32); 20] = self.bit_array.iter().zip(other.bit_array.iter()).collect::<Vec<_>>().try_into().unwrap();
            let zipped_bit_array: [(&i32, &i32); 20] = self.bit_array.iter().zip(other.bit_array.iter()).collect::<Vec<(&i32, &i32)>>().try_into().unwrap();

            let zipped_bit_array  = self.bit_array.iter().zip(other.bit_array.iter()).collect::<Vec<_>>();
            let mut new_bit_array = vec![];

            for item in zipped_bit_array {
                new_bit_array.push(item.0 | item.1);
            }
         */

        /*
            let new_bit_array = self.bit_array.iter()
                .zip(other.bit_array.iter())
                .map(|item| item.0 | item.1)
                .collect::<Vec<_>>();
         */

        let new_bit_array = self.bit_array.iter()
            .zip(other.bit_array.iter())
            .map(|(x, y)| x | y)
            .collect::<Vec<_>>();

        let mut result = Self::new(self.size, self.number_hashes);
        result.bit_array = new_bit_array;

        result
    }

    pub fn intersection(&self, other: BloomFilter) -> Self {
        if self.size != other.size || self.number_hashes != other.number_hashes {
            panic!("Both filters must have the same size and hash count")
        }

        Self {
            size: self.size,
            number_hashes: self.number_hashes,
            salt: "".to_string(),
            bit_array: self.bit_array.iter()
                .zip(other.bit_array.iter())
                .map(|(x, y)| x & y)
                .collect::<Vec<_>>(),
        }
    }

}

//
impl Sha1Hash for BloomFilter {

    fn sha1_hash(&self, value: &str) -> String {
        let mut hasher = Sha1::new();
        hasher.update(value);

        hex::encode(hasher.finalize())
    }

}

//
#[allow(dead_code)]
fn type_of<T>(_: &T) -> String {
    return format!("{}", std::any::type_name::<T>())
}

fn main() {
    let coffees = [
        "Iced Coffee",
        "Iced Coffee with Milk",
        "Espresso",
        "Espresso Macchiato",
        "Flat White",
        "Latte Macchiato",
        "Cappuccino",
        "Mocha",
    ];

    let mut bloom = BloomFilter::new(20, 2);
    for drink in coffees.iter() {
        bloom.add(drink);
        println!("{:?}", bloom.bit_array);
    }

    println!("{}", "---Experiment #1---");
    println!("{}", bloom.lookup("Flat White"));
    println!("{}", bloom.lookup("Americano"));
    println!("{}", bloom.estimate_dataset_size());

    let more_coffees = [
        "Iced Espresso",
        "Flat White",
        "Cappuccino",
        "Frappuccino",
        "Latte",
    ];

    let mut bloom2 = BloomFilter::new(20, 2);
    for i in 0..more_coffees.len() {
        bloom2.add(more_coffees[i]);
    }

    let bloom3 = bloom2.union(bloom.clone());
    println!("{}", "---Experiment #2---");
    println!("{}", bloom3.lookup("Mocha"));
    println!("{}", bloom3.lookup("Frappuccino"));
    println!("{}", bloom3.estimate_dataset_size());

    let bloom4 = bloom2.intersection(bloom.clone());
    println!("{}", "---Experiment #3---");
    println!("{}", bloom4.lookup("Mocha"));
    println!("{}", bloom4.lookup("Flat White"));
    println!("{}", bloom4.estimate_dataset_size());
}
